job-list: support queue specific stats

chu11 opened the pull request:
per #4604, get queue specific job stats.  in `flux-jobs`, the `--queue` option can now also select which queue `--stats` outputs.

chu11 commented:
re-pushed, fixed up && chain in tests, valgrind memleak find, and python linting

garlick commented on src/modules/job-list/stats.c:
```diff
@@ -157,12 +210,69 @@ static json_t *stats_encode (struct job_stats *stats)
+
+    if (name) {
+        json_t *no = json_string (name);
+        if (!no || json_object_set (o, "name", no) < 0) {
```
This leaks `no` on success.  I think you meant `json_object_set_new()`?

garlick commented on src/modules/job-list/stats.c:
```diff
@@ -157,12 +210,69 @@ static json_t *stats_encode (struct job_stats *stats)
+        const char *name = zhashx_cursor (statsctx->queue_stats);
+        json_t *qo = stats_encode (stats, name);
+        if (!qo) {
+            json_decref (queues);
```
json_decref() clobbers errno

garlick commented on src/modules/job-list/stats.c:
```diff
@@ -157,12 +210,69 @@ static json_t *stats_encode (struct job_stats *stats)
     return o;
 }
 
+static int queue_stats_encode (struct job_stats_ctx *statsctx,
+                               json_t *o)
```
The function name does not suggest that the stats are going to be inserted in `o`.  Maybe have the caller do that?

garlick commented on src/modules/job-list/stats.c:
```diff
@@ -103,9 +151,14 @@ static void stats_purge (struct job_stats *stats, struct job *job)
 {
+    struct job_stats *stats;
+
     assert (job->state == FLUX_JOB_STATE_INACTIVE);
```
An assertion here kills the rank 0 broker.  Appropriate in this case?

chu11 replied:
i think that assert was meant as a double check that this is only called on inactive jobs, but I see

```
    if (flux_event_unpack (msg, NULL, "{s:o}", "jobs", &jobs) < 0)                                                                          
        flux_log_error (h, "job-purge-inactive message");                                                                                   
    json_array_foreach (jobs, index, entry) {                                                                                               
        flux_jobid_t id = json_integer_value (entry);                                                                                       
        struct job *job;                                                                                                                    
                                                                                                                                            
        if ((job = zhashx_lookup (ctx->jsctx->index, &id))) {                                                                               
            job_stats_purge (ctx->jsctx->statsctx, job);                                                                                    
            if (job->list_handle)                                                                                                           
                zlistx_delete (ctx->jsctx->inactive, job->list_handle);                                                                     
            zhashx_delete (ctx->jsctx->index, &id);                                                                                         
            count++;                                                                                                                        
        }                                                                                                                                   
    }                                                          
```

that we do not verify this fact.  it trusts the fact that every job id passed to it is inactive.  We should double check that within the loop.

I'll create an issue, b/c this is independent of this PR.

garlick commented on doc/man1/flux-jobs.rst:
```diff
@@ -74,23 +74,27 @@ OPTIONS
-   Output a summary of global job statistics before the header.
-   May be useful in conjunction with utilities like
-   :linux:man1:`watch`, e.g.::
+   Output a summary of job statistics before the header.  By default will
+   output global statistics.  If ``--queue`` is specified, will output
+   statistics for the specified queue.  May be useful in conjunction with
+   utilities like :linux:man1:`watch`, e.g.::
```
Maybe s/will output/show/?

garlick commented on doc/man1/flux-jobs.rst:
```diff
@@ -74,23 +74,27 @@ OPTIONS
-   Output a summary of global job statistics and exit.
-   ``flux jobs`` will exit with non-zero exit status with ``--stats-only``
-   if there are no active jobs. This allows the following loop to work::
+   Output a summary of job statistics and exit.  By default will output
+   global statistics.  If ``--queue`` is specified, will output
+   statistics for the specified queue.  ``flux jobs`` will exit with
+   non-zero exit status with ``--stats-only`` if there are no active
+   jobs. This allows the following loop to work::
```
s/will output/show/

garlick reviewed (commented):
I tried running this branch on my test system that has two queues (batch and debug).  Initially, I ran no new jobs and got no per queue stats reported even though there were inactive jobs in the KVS.  Then I ran a job in the batch queue that happened to fail, and got these stats
```
 garlick@picl0:~/proj/flux-core$ flux jobs --stats --queue=batch
0 running, -1 completed, 1 failed, 0 pending
       JOBID USER     NAME       ST NTASKS NNODES     TIME INFO
```
so something's a little off there.

chu11 commented:
> so something's a little off there.

hmmm.  maybe a bug related to restarting of a flux instance, which I may have not added a test for.  Thanks!

chu11 commented on src/modules/job-list/stats.c:
```diff
@@ -18,6 +18,24 @@
+void job_stats_ctx_destroy (struct job_stats_ctx *statsctx)
+{
+    if (statsctx)
+        free (statsctx);
```
ahh, the `if (statsctx)` was a little bit "with future expectations of need", should be in a later commit instead.

chu11 commented:
re-pushed, fixing up the issues @garlick found above

fixed queue specific stats on reload of the job-list module and added some tests & extra coverage for it

However, I was not able to figure out how

> 0 running, -1 completed, 1 failed, 0 pending

occurred though.  It's perplexing.  Basically in the python code the number of completed jobs is:

`self.successful = self.inactive - self.failed`

suggesting inactive job count was < the failed job count to get the `-1` output.  Which I can't see a way in which that could happen.  

My only guess is it could have been a counting issue with a job that was active when the job-list module was reloaded and one of the counters may have been under-counted.  But I don't think that's what happened here.

garlick commented:
> My only guess is it could have been a counting issue with a job that was active when the job-list module was reloaded and one of the counters may have been under-counted. But I don't think that's what happened here.

This might be unrelated to this PR.  Watch this:
```
 garlick@picl0:~$ flux jobs --stats-only
0 running, 87 completed, 13 failed, 0 pending
 garlick@picl0:~$ flux mini run -q debug /bin/false
flux-job: task(s) exited with exit code 1
 garlick@picl0:~$ flux jobs --stats-only
0 running, 86 completed, 14 failed, 0 pending
```
Somehow the completed count dropped.  This is on master right after #4687 (emoji) was merged.

garlick reviewed (approved):
Just one minor comment.  I tried this again on my system with queues configured and it seems to behave as expected!

chu11 commented on src/modules/job-list/stats.c:
```diff
@@ -18,19 +18,62 @@
+        if (!(stats = calloc (1, sizeof (*stats))))
+            return NULL;
+        if (zhashx_insert (statsctx->queue_stats, job->queue, stats) < 0) {
+            flux_log_error (statsctx->h, "%s: zhashx_insert", __FUNCTION__);
```
oops, it seems I do that all over the place and with `zlistx` functions too.  I'll leave this here for now, do a bigger cleanup in a followup PR.
